app/utils/webhook.js

```diff
@@ -0,0 +1,177 @@
+/**
+* Copyright 2019 IBM Corp. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+const uuid = require('uuid');
+const request = require('request-promise-native');
+const objectPath = require('object-path');
+const { URL } = require('url');
+
+const WEBHOOK_TRIGGER_IMAGE = 'image';
+const WEBHOOK_TRIGGER_CLUSTER = 'cluster';
+
+// fireWebhook - private method for calling a web hook
+const fireWebhook = async (webhook, postData, req) => {
+  var success = true;
+  var webhookId = webhook.id;
+  var url = webhook.service_url;
+  var hasError = false;
+  var resp = '';
+  postData.request_id = uuid();
+  const options = {
+    method: 'POST',
+    uri: url,
+    body: postData,
+    json: true
+  };
+  req.log.debug(options, 'fireWebhook');
+
+  try {
+    req.log.debug({ org_id: req.org._id, options: options }, 'POSTing webhook');
+    resp = await request(options);
+  }
+  catch (e) {
+    success = false;
+    hasError = true;
+    resp = e;
+    req.log.error({ err: e.message, url }, 'Error POSTing webhook');
+  }
+  var logObj = {
+    webhook_id: webhookId,
+    org_id: req.org._id,
+    kind: webhook.kind,
+    req: {
+      url: url,
+      payload: JSON.stringify(postData),
+    },
+    res: JSON.stringify(resp),
+    hasError,
+    created: new Date(),
+  };
+  const WebhookLogs = req.db.collection('webhookLogs');
+  WebhookLogs.insert(logObj);
+  return success;
+};
+
+// processWebhooks - private method for filtering and executing web hooks
+const processWebhooks = async (webhooks, postData, resourceObj, req) => {
+  req.log.debug(webhooks, 'processWebhooks');
+  var success = true;
+  await Promise.all(webhooks.map(async (webhook) => {
+    // Test if optional filter
+    var match = true;
+    if (webhook.filter) {
+      var field = objectPath.get(resourceObj, webhook.field);
+      if (!field.match(webhook.filter)) {
+        match = false;
+      }
+    }
+    if (match) {
+      postData.webhook_id = webhook._id;
+      if (!await fireWebhook(webhook, postData, req)) {
+        success = false;
+      }
```

Comment:
Alternate impl:
``` javascript 
success = !await fireWebhook(webhook, postData, req);
```
Do we want the ! of fireWebhook(), or do we want to send its result along?

Reply:
changed to success = await fireWebhook.   we want to know if it failed.

Reply:
done